### PR TITLE
Migrate scripted metric aggregation scripts to ScriptContext design

### PR DESCRIPTION
**Pull request by rationull**
This is the start to an implementation of new script contexts for scripted metric agg scripts as outlined in #29328. I don't think this is ready to merge yet but wanted to submit it for feedback on some specific questions and also a sanity check on general direction.

General questions (I will comment in the diff for more specific points):

1. I've used "agg" and "aggs" as the variable names, as direct translation from the old params._agg and _aggs. Should these have leading underscores like "_score" or no underscore like "params"? Any other suggestion for names? RESOLUTION: Went with "state"/"states"
2. Detailed backwards compatibility strategy still needs to be fleshed out (specifically when and how deprecation warnings need to be emitted). RESOLUTION: Deprecation warning and system property to disable deprecated behavior will come in a subsequent PR.
3. I haven't done any work yet to support the new contexts in expression scripts. Obviously this is required but didn't want to get too far ahead of the general sanity check. RESOLUTION: Looks like no work necessary here since expression scripts can't effectively be used for scripted metric aggregations.
4. I haven't cleaned up the old AGGS_CONTEXT in SearchScript and ExecutableScript which presumably are not needed after this. Though they are still used -- for reduce scripts in bucket aggregations and in ExecutableScriptHeuristic, in ValueSourceConfig, and in some tests. I suspect most or all of these should actually be replaced with different contexts.
5. I realize the commit comment could use more detail/background as well.
6. Documentation changes. RESOLUTION: Done as part of this PR

### CODE REVIEW

File: modules/lang-painless/src/test/java/org/elasticsearch/painless/MetricAggScriptsTests.java
Comment (rationull) [inline modules/lang-painless/src/test/java/org/elasticsearch/painless/MetricAggScriptsTests.java]: I added these tests just for convenient sanity checking that the new contexts were working. But perhaps they're redundant with more general Painless tests of script parameter access functionality.

File: server/src/main/java/org/elasticsearch/script/MetricAggScripts.java
Comment (rationull) [inline server/src/main/java/org/elasticsearch/script/MetricAggScripts.java]: Is there any reason that agg should be provided via execute() parameter instead of via an accessor like this?

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregatorTests.java
Comment (rationull) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregatorTests.java]: ScoreAccessor here doesn't work with the new contexts. I don't see why this use should be coupled specifically to ScoreAccessor.

File: server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricIT.java
Comment (rationull) [inline server/src/test/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricIT.java]: Added a new test here for now to get some specific coverage easily. I think what actually makes sense here is to run all the test cases here against both old param and new context variable scripts, while both are implemented.

**Comment by javanna** [comment]
@elastic/es-search-aggs thoughts on this one?

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java
Comment (colings86) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java]: Is this needed? I thought that the reason to use script contexts was to that we then wouldn't need to put the `_aggs` variable in the params because it is passed directly to the script?

File: server/src/main/java/org/elasticsearch/script/MetricAggScripts.java
Comment (colings86) [inline server/src/main/java/org/elasticsearch/script/MetricAggScripts.java]: Can we call this `ScriptedMetricAggContexts` instead since "metric agg" is a term that includes all leaf aggregation types not just the `scripted_metric` type?

**Review by colings86 — DISMISSED** [review]
@rationull Sorry for the delay in reviewing this. I left some comments but I think it looks good in general. I would like @rjernst to also look at it as well before we merge

**Comment by colings86** [comment]
@elasticmachine please test this

**Comment by colings86** [comment]
@elasticmachine test this please

File: server/src/main/java/org/elasticsearch/script/MetricAggScripts.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/script/MetricAggScripts.java]: Absolutely -- that is much clearer. I will make that change ASAP.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java]: This (along with analogous code for `_agg` in ScriptedMetricAggregatorFactory.java and some of the integration tests I did not modify) is needed if we want to retain backwards compatibility with the old way for one ES version, letting people migrate to the context variables while both still work.

If a clean break is acceptable then this would be removed and some of the tests could be cleaned up. But I would assume we want some compatibility leniency and possibly some warning messages or other notification (which I have not implemented yet -- item 2 on my list in the first comment on this PR).

Comment (colings86) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java]: Ok, understood. I'm +1 on keeping this for backwards compatibility, thanks for explaining. Maybe we could add a comment just noting that its there for backwards compatibility to avoid someone removing it thinking its just left over?

Comment (rationull) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java]: Good call. I will add a comment here and in other relevant spots.

**Comment by rationull** [comment]
Master is presently failing in CI but I merged it into my branch anyway so at least any further builds should fail consistently with it until it’s fixed and this branch can be updated again.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/InternalScriptedMetric.java]: Glad you brought this up -- when looking for spots that needed a comment I found a leftover usage of `params.get("_agg")` in ScriptedMetricAggregator.java that needed to be replaced with a direct usage of the agg object. No functional issue but it would've had to get cleaned up later.

**Comment by rationull** [comment]
@rjernst note I made some comments when creating this PR which are still relevant but are hidden as outdated after my last minor cleanup commit. Specifically comments about naming and usage of context variables vs `execute()` parameters.

**Comment by rationull** [comment]
@rjernst any feedback on this?

File: server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java]: I would name this MetricAggScripts so that the inner classes read more naturally, eg `MetricAggScripts.Init`.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java]: This will cause the score to always be calculated, even when not being used. Instead, in SearchScript, we have the `get_score()` method, which painless converts to a `_score` variable and is only invoked when it is actually used. So, the MapScript should have `setScorer`

File: server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java]: I think this base class could be removed by passing the arguments directly to `execute()`?

**Review by rjernst — COMMENTED** [review]
Sorry for the long delay in responding. In general, I wonder if the "agg" name should continue to be used, or if something better like "state" could make it more clear what the purpose of the object is for?

Also, in order to have a clear deprecation path for using params._agg, I would control adding that with a system property, which emits a deprecation warning when read the first time if it is enabled, but that could be done in a followup.

File: server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java]: That's what I had originally, actually. I abandoned e.g. `Init` in favor of `InitScript` because naming one of the classes `Map` ends up requiring fully qualifying java.util.Map in a bunch of spots which is pretty annoying (although really isn't a dealbreaker since it could be limited to this file). And past that, @colings86 suggestion that "metric agg" as a term is more general than this feature made sense to me. Admittedly the current situation makes the outer class kind of unnecessary since references to the scripts end up using long names anyway. This could be replaced with a package (or not even) and names like `ScriptedMetricAggInitScript` or just `MetricAggInitScript` which may be more palatable since it avoids a more general name without "script" in it.

TBH I might lean toward the latter; using a class as a small namespace in this context is a little clunky anyway IMO (even if it's the best Java can do). I defer to what you two think makes the most sense in context. I will await comment from @rjernst and @colings86 to settle this one.

File: server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java]: I did have a general question (hidden/outdated now in this PR) about `execute()` args vs accessors. Is there a general intent in the script context design for which mechanism any given script context variable should use? My read was that "ambient" or earlier-bound context should be via accessors and more execution specific or later-bound context should be via `execute()` args.

In this case specifically both `agg` and `params` are conveniently available at the time the script contexts are created, but less conveniently (in some cases) available when the script is executed, so I opted to use accessors instead. They could be passed down the chain of course if that design is preferable for some reason.

The base class could be removed regardless of that though, IMO just a matter of style taste. I have no love for extraneous classes or deep inheritance so happy to remove it if you feel it just complicates things.

**Comment by rationull** [comment]
@rjernst I like the idea of a more general name like `state`. `agg` still seems like a valid name since the object represents the aggregation in progress. I think ideally it would be named something more descriptive but it's hard to balance that with also keeping it short. Even `result` could be a good option. This deserves some more thought. 

re the deprecation warning: to make sure I understand, are you proposing that the old `params._agg` would not even be present by default, and thus anyone migrating code that uses it would have to enable this new `useLegacyAggScriptParameter` (or whatever) setting? So that way the first hint (aside from release notes) would be a script failure, and the second hint would be the docs + the warning that appears when you set the flag? That sounds like good fail-fast behavior.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java]: `get_score()` should work; if it doesn't, it's a bug. @jdconrad Can you check this?

File: server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java]: I had thought the params were setup next to the execution, but I see now the extra level of indirection in the aggs classes. I'm fine with the way you have it in that case.

More generally, the args and accessors are mostly interchangeable, for the reasons you stated. Args are necessary when the value will change per execution.

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java]: Sorry, I was mistaken here.. another quick hack test just now shows variables that start with underscores do work. Not sure what test lead me to the conclusion before that it didn't. That should make this an easy fix. No need to mobilize @jdconrad :)

**Comment by rationull** [comment]
Got it. I think we were on the same page other than the default state of the system property. I didn't realize this would likely end up in 6.x; obviously we wouldn't want to make it break by default in that case.

@rjernst I will work on updates per your suggestions and probably have an update for the PR either this weekend or in a couple weeks, time permitting. Thanks!

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/scripted/ScriptedMetricAggregator.java]: My last commit resolved this.

**Comment by rationull** [comment]
@rjernst @colings86 I believe the following are the remaining issues, each noted with how I think it needs to proceed. Input is appreciated:

1. `agg`/`aggs` naming: Ryan suggested `state` (and presumably `states`). I might suggest `result`/`results` as well although that could be misleading since you can't assign the identifier itself to a new value and have that percolate across the other scripts. Or we can keep the existing naming. @colings86 input on this?
2. Deprecation warning with associated flag. @rjernst suggested this could be done in a followup PR. That sounds good to me, to get this one wrapped up. But I'm happy to do that under this PR if you're more comfortable with that.
3. Documentation changes. I'm not sure what the process is wrt tech writing but I've been assuming I'll need to work on that. I've also been assuming that'd be in a separate PR but again happy to do it here if needed.
4. Other cleanup that may be possible (e.g. other `AGGS_CONTEXT` cleanup). I think this is best split into separate issues and separate PRs since it is mostly related to features other than scripted metric aggregations. I can note this in #29328 and either pursue the related changes or not depending on their value and my available time.

File: server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java
Comment (rationull) [reply server/src/main/java/org/elasticsearch/script/ScriptedMetricAggContexts.java]: @rjernst I'm inclined to leave this as-is because personally I think MapScript name qualification is more annoying than the ScriptedMetricAggContexts.*Script naming, and on further reflection I do think having them all in one file is better than splitting them apart. Let me know if you are firm in a different opinion on this though.

**Comment by colings86** [comment]
> 1. agg/aggs naming: Ryan suggested state (and presumably states). I might suggest result/results as well although that could be misleading since you can't assign the identifier itself to a new value and have that percolate across the other scripts. Or we can keep the existing naming. @colings86 input on this?

I don't feel too strongly on this so any of the suggestions would be fine with me, although I do lean towards either `agg`/`aggs` or `result`/`results`

> 3. Documentation changes. I'm not sure what the process is wrt tech writing but I've been assuming I'll need to work on that. I've also been assuming that'd be in a separate PR but again happy to do it here if needed.

We generally make documentation changes int he same PR as the code change so they are together so I would prefer if we made the necessary doc changes in this PR

**Comment by rjernst** [comment]
> I don't feel too strongly on this so any of the suggestions would be fine with me, although I do lean towards either agg/aggs or result/results

I don't like `agg` because it is completely non descriptive. It does not convey what is stored in it (the state across script invocations). I don't like `result` for the reason @rationull mentioned.  I suggested `state` because I thought it was simple. If you both feel it is too generic, maybe `metricState` is better?

**Comment by rationull** [comment]
Well, I haven't come up with anything better. I would lean toward `state` rather than `metricState` since I think for a scripting use case shorter length outweighs more descriptiveness. `state` is descriptive in this case even if it's a very general term.

**Comment by rationull** [comment]
Added a first attempt at the required doc changes. I have a separate small branch for the deprecation warning with additional doc changes related to the proposed system property which I can keep as a separate small PR after this gets merged or I can tack on here (although this one has gone on for quite a while already).

I can squash all the changes and write a better commit comment as well but don't want to do that until both @rjernst and @colings86 consider all the PR comments resolved.

Guidance is appreciated but AFAIK this is just waiting on a conclusion on `agg` vs `state` naming, final approval, and a passing test run (although the master branch seems to be failing quite a bit lately in CI).

**Comment by colings86** [comment]
I'm ok with using `state`/`states`

**Comment by colings86** [comment]
jenkins test this please

**Comment by rationull** [comment]
Updated the code and the docs with the `state`/`states` rename, and merged in master again to resolve a doc conflict.

I'm not aware of any other issues that need to be resolved here if there is no feedback on the doc changes or any new comments on the code changes. @colings86 your review is still showing as "changes requested" but my understanding is this is just outdated from the first review.

**Comment by rationull** [comment]
@colings86 @rjernst the rename is done and all the doc updates that I know about are made.

**Comment by colings86** [comment]
@rationull thanks for updating the PR. I've dismissed my review since its out of date now. I see there are some merge conflicts now, could you update the branch with the latest master and push. Then I'll set off a CI build and do another review

**Comment by rationull** [comment]
@colings86 yep my timing on my last comment was a bit unfortunate as another small tweak in a different PR was merged a few hours later :) The conflicts are expected. I will resolve this after work today.

**Comment by rationull** [comment]
@colings86 conflicts resolved.

**Comment by colings86** [comment]
@elasticmachine jenkins test this please

**Review by colings86 — COMMENTED** [review]
@rationull this looks good but I would like @rjernst to take a final look to make sure he is happy with the script specific changes before we merge.

**Comment by rationull** [comment]
Awesome, thanks @colings86 for the review.

**Comment by rjernst** [comment]
I'm ok with the script specific changes. While I think long term we should investigate using accumulators (instead of an arbitrary map), this will depend a lot on some painless optimizations that need to be made in order to fully utilize the performance improvements accumulators can offer (note, that also may allow scripted metric aggs to be on par with java based metric aggs).

**Comment by colings86** [comment]
Thanks @rjernst for taking another look.

@rationull could I ask you to update the branch with the latest master on more time, then I'll get the build to run and we can merge this.

**Comment by colings86** [comment]
@elasticmachine test this please

**Comment by colings86** [comment]
@rationull Thanks for submitting this PR, I think it will be a really positive change for the scripted_metric aggregation.

**Comment by rationull** [comment]
@colings86 no problem, thanks for the reviews! I have a couple of backwards compatibility/deprecation followup changes that I'll submit in the next few days. I will update the issue with details as well.

**Comment by duxy-ios** [comment]
Thanks rationull and colings86  ,  When can we use this wonderful improvement
@colings86  ，We have an urgent need for it

**Comment by rjernst** [comment]
@duxy-ios This will be in 6.4.0.

**Pull request merged**
